ggml/src/ggml-aarch64.c

```diff
@@ -545,73 +557,67 @@ void ggml_gemv_q4_0_4x4_q8_0(int n, float * restrict s, size_t bs, const void *
     UNUSED(ncols_interleaved);
     UNUSED(blocklen);
 
-#if defined(__ARM_FEATURE_SVE)
-    if (ggml_sve_cnt_b == QK8_0) {
-        GGML_ASSERT(!(ggml_cpu_has_sve() && (ggml_sve_cnt_b == QK8_0)) &&
-                    "__ARM_FEATURE_SVE defined, use the Q4_0_8_8 quantization format for optimal performance");
+#if ! ((defined(_MSC_VER)) && ! defined(__clang__)) && defined(__aarch64__) && defined(__ARM_NEON)
+    if (ggml_cpu_has_neon()) {
```

Comment:
I'm wondering if these function calls are properly inlined. AFAIK with LTO enabled, they should be, but maybe it's better if instead of relying on the compilation to do it for us, we can read the value once into a static variable and check that variable from the on.

The function call overhead is probably negligible, but still. What do you think?

Reply:
@ggerganov Thanks for the review and sorry for the late response as I was on vacation. I'll look into your suggestion.

Reply:
One possible way to reduce the overhead of function calls such as ggml_cpu_has_neon() is to cache the results using static variables, as you suggested:

```
static bool neon_support_checked = false;
static bool neon_supported = false;
```

To prevent race conditions, the values of neon_support_checked and neon_supported need to be protected by a mutex or atomic operations:

```
static pthread_mutex_t neon_check_mutex = PTHREAD_MUTEX_INITIALIZER;

inline bool is_neon_supported() {
    pthread_mutex_lock(&neon_check_mutex);

    // Check only if not already checked
    if (!neon_support_checked) {
        neon_supported = ggml_cpu_has_neon();
        neon_support_checked = true;
    }

    pthread_mutex_unlock(&neon_check_mutex);  // Release the lock
    return neon_supported;
}
```
By marking is_neon_supported() as inline, we may reduce the function call overhead. However, since the function still contains a mutex, the overall performance improvement could be limited due to the cost of acquiring and releasing the lock.

Another possible optimization might be to inline the ggml_cpu_has_neon() function itself, although that might be beyond the scope of this PR.

I'm interested to hear what you think, and I'm happy to consider any suggestions or improvements you may have.

Reply:
Nah, mutex is an overkill. Let's merge it as it is. You can easily check if the function calls add any overhead by replacing with `if (true)`. If they do, you can try to find a way to do thread-safe static init without synchronization primitives.

Reply:
I’ve done some overhead tests for the function calls and didn’t observe any statistically significant performance differences. Given that, I’m fine with merging the code as is after I rebase it. Please let me know if you think I should do anything differently.

Reply:
Ok, let's merge after rebase.